platform/local/dnsmasq.go

```diff
@@ -147,6 +207,85 @@ func newSegment(s byte) (*Segment, error) {
 		return nil, fmt.Errorf("LinkSetUp() failed: %v", err)
 	}
 
+	// we first create a virtual ethernet pair in current network namespace
+	// based on the current process ID
+	pid := os.Getpid()
+	pair := generateVethPair(pid)
+
+	peer0 := strings.Split(pair[0], "|")
+	peer1 := strings.Split(pair[1], "|")
+
+	attr := netlink.NewLinkAttrs()
+	attr.Name = peer0[0]
+	veth := &netlink.Veth{
+		PeerName:  peer1[0],
+		LinkAttrs: attr,
+	}
+
+	if err := setupLink(veth, peer0[1], true); err != nil {
+		return nil, fmt.Errorf("unable to set up link: %w", err)
+	}
+
+	peer, err := netlink.LinkByName(peer1[0])
+	if err != nil {
+		return nil, fmt.Errorf("unable to get link by name: %w", err)
+	}
+
+	// move to root network namespace
+	if err := netlink.LinkSetNsPid(peer, 1); err != nil {
+		return nil, fmt.Errorf("unable to set link into a new ns: %w\n", err)
+	}
+
+	gtw, _, err := net.ParseCIDR(pair[2])
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse CIDR address: %w", err)
+	}
+
+	_, dst, err := net.ParseCIDR("0.0.0.0/0")
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse CIDR address: %w", err)
+	}
+
+	if err := netlink.RouteAdd(&netlink.Route{
+		Scope:     netlink.SCOPE_UNIVERSE,
+		Dst:       dst,
+		LinkIndex: veth.Attrs().Index,
+		Gw:        gtw,
+	}); err != nil {
+		return nil, fmt.Errorf("unable to add default route: %w", err)
+	}
+
+	if err := configureNAT(peer0[0]); err != nil {
+		return nil, fmt.Errorf("unable to configure NAT: %w", err)
+	}
+
+	root, err := netns.GetFromPid(1)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get NS from PID 1: %w", err)
+	}
+
+	if err := netns.Set(root); err != nil {
```

Comment:
Seems this is not guaranteed due to how Go routines are scheduled and suspendable. Instead, the original version of `.Enter()` should be used to switch and go back.